route-snapper/src/tests.rs

```diff
@@ -0,0 +1,210 @@
+    // setRouteConfig is a WASM API awkward to call; just set directly
+    snapper.router.config.extend_route = true;
+
+    // Add many waypoints
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT3);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2, WAYPT3]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT4);
+    snapper.on_click();
+    assert_eq!(
+        snapper.route.waypoints,
+        vec![WAYPT1, WAYPT2, WAYPT3, WAYPT4]
+    );
+
+    // Clicking an existing waypoint will still delete it
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT3, WAYPT4]);
+
+    // TODO Drag
+}
+
+#[test]
+fn test_route_dont_extend() {
+    let map_bytes = std::fs::read("../examples/southwark.bin").unwrap();
+    let mut snapper = JsRouteSnapper::new(&map_bytes).unwrap();
+
+    // The first two waypoints work normally
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    // But then we can't add another one
+    optionally_mouseover_waypt(&mut snapper, WAYPT3);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    // Delete the first waypoint
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT2]);
+
+    // Then add a different endpoint
+    must_mouseover_waypt(&mut snapper, WAYPT4);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT2, WAYPT4]);
+
+    // TODO Drag
+}
+
+#[test]
+fn test_area() {
+    let map_bytes = std::fs::read("../examples/southwark.bin").unwrap();
+    let mut snapper = JsRouteSnapper::new(&map_bytes).unwrap();
+    snapper.set_area_mode();
+
+    // The first two points just make a line
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    // The third creates the polygon, and the first and last waypoints are now the same
+    must_mouseover_waypt(&mut snapper, WAYPT3);
+    snapper.on_click();
+    assert_eq!(
+        snapper.route.waypoints,
+        vec![WAYPT1, WAYPT2, WAYPT3, WAYPT1]
+    );
+
+    // Can't delete an intermediate waypoint if there aren't enough
+    // TODO Actually not true; decide whether the behavior now is what we want or not
```

Comment:
I had a play around. This could definitely use some thought, current behaviour is a little surprising since deleting the 3rd waypoint doesn't return you to the previous state. That said idk how I would want it to work